ch08performance/030misc-libs.ipynb.py

```diff
@@ -0,0 +1,385 @@
+# optimizations than JIT, but AOT compilers produce huge binaries and that too only
+# for specific platforms. JIT compilers have limited optimization routines but they
+# produce small and platform independent binaries.
+# 
+# JIT and AOT compilers for Python include, but are not limited to -
+#  - [Numba](http://numba.pydata.org): a Just-In-Time Compiler for Numerical Functions in Python
+#  - [mypyc](https://mypyc.readthedocs.io/en/latest/introduction.html): compiles Python modules to C extensions using standard Python type hints
+#  - [JAX](https://jax.readthedocs.io/en/latest/jit-compilation.html): a Python library for accelerator-oriented array computation and program transformation
+# 
+# Although all of them were built to speed-up Python code, each one of them is a bit
+# different from each other when it comes to their use-case. We will specifically look into
+# Numba in this lesson.
+# 
+# ### Numba
+# 
+# Numba is an open source JIT compiler that translates a subset of Python and NumPy code
+# into fast machine code. The good thing about Numba is that it works on plain Python
+# loops and one doesn't need to configure any compilers manually to JIT compile Python code.
+# Another good thing is that it understands NumPy natively, but as detailed in its
+# documentation, it only understands a subset of NumPy functionalities.
+# 
+# Numba provides users with an easy to use function decorator - `jit`. Let's start by
+# importing the libraries we will use in this lesson.
+#
+
+# %%
+import math
+
+from numba import jit, njit
+import numpy as np
+
+# %% [markdown]
+# We can mark a function to be JIT compiled by simply decorating it
+# with numba's `@jit`. The decorator takes a `nopython` argument that tells Numba to
+# enter the compilation mode and not fall back to usual Python execution.
+
+# %%
+def f(x):
+    return np.sqrt(x)
+
+@jit(nopython=True)
+def jit_f(x):
+    return np.sqrt(x)
+
+# %% [markdown]
+# It looks like the `jit` decorator should make our Numba compile our function
+# and make it much faster than the non-jit version. Let's test this out.
+# 
+# Note that the first function call is not included while timing because that
+# is where Numba compiles the function. The compilation at runtime is called
+# just in time compilation and resultant binaries are cached for the future
+# runs. 
+
+# %%
+data = np.random.uniform(low=0.0, high=100.0, size=1_000)
+
+# %%
+# %%timeit
+f(data)
+
+# %%
+_ = jit_f(data)
+
+# %%
+# %%timeit
+jit_f(data)
+
+# %% [markdown]
+# Surprisingly, the JITted function was slower than plain Python and NumPy
+# implementation! Why did this happen? Numba does not provide valuable performance
+# gains over pure Python or NumPy code for simple operations and small dataset.
+# The JITted function turned out to be slower than the non-JIT implementation
+# because of the compilation overhead. Let's try increasing the size of our
+# data and perform a non-NumPy list comprehension on the data.
+# 
```

Comment:
Yes, I was just trying to time the running time and not the compile time. Should I not compile the function before timing it?

Reply:
I think, in that case, it doesn't produce any difference, because as you said, the optimisation is not going to be much better than what numpy already does. We could leave it as it is and explain that for measuring JIT performance is normally done in that way. Though ideally, we should also measure the compilation to show that overhead.
So, I'd say, in all the jit examples, run it once with:
```
%timeit -n 1 -r 1
jit_function(data) # compilation and run
```
and then
```
%timeit
jit_function(data) # just run
```
This way we ensure that the compilation timing is calculated separately and don't affect the average (though in these examples is probably spread out). We need to note, however, that the result from the compilation run could be very noisy and could give a higher than real value (I think we mention that in the previous lesson when introducing `timeit`).

Reply:
Thank you! Working on this right now 🚀